### PR TITLE
[improvement](bitshuffle)Enable avx512 support in bitshuffle for performance boost

### PR DESCRIPTION
**Pull request by HackToday**
Signed-off-by: Wu, Kaiqiang <kaiqiang.wu@intel.com>
Co-authored-by: vesslanjin <jun.i.jin@intel.com>

# Proposed changes

Issue Number: close #xxx

## Problem summary

As AVX512 is available in most modern processors, it is good to use them if have performance boost.
In latest bitshuffle, AVX512 have been added. We could make it integrated in doris for AVX512 case.

Tested with master branch,  queries(SSB query q1.1.sql~q4.3.sql total 13 queries) can be boost from 1.4%~3.2%. (use run-ssb-queries.sh 5 times,  each time with 100 iterations.)

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/gutil/cpu.cc [hunk 1]
```diff
@@ -253,6 +256,7 @@ void CPU::Initialize() {
 #endif
 }
 CPU::IntelMicroArchitecture CPU::GetIntelMicroArchitecture() const {
+    if (has_avx512()) return AVX512;
```
[the remark this answers is not in the document]
Comment (HackToday) [reply be/src/gutil/cpu.cc:259]: It seems old code not use braces. so not sure if needed to change all if case?

[the remark this answers is not in the document]
Comment (adonis0147) [reply be/src/gutil/cpu.cc:259]: > It seems old code not use braces. so not sure if needed to change all if case?

This message is a warning, you can ignore it.

File: be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp [hunk 2]
```diff
@@ -34,6 +34,15 @@
...
 
+#undef BITSHUFFLE_H
+#define bshuf_compress_lz4_bound bshuf_compress_lz4_bound_avx512
+#define bshuf_compress_lz4 bshuf_compress_lz4_avx512
+#define bshuf_decompress_lz4 bshuf_decompress_lz4_avx512
+#include <bitshuffle/bitshuffle.h> // NOLINT(*)
+#undef bshuf_compress_lz4_bound
+#undef bshuf_compress_lz4
+#undef bshuf_decompress_lz4
+
```
Comment (adonis0147) [inline be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp:45]: Hi @HackToday , these lines look tricky. Why should we do this?

**Comment by adonis0147** [comment]
> @morningman Could you help make the workflow run ? It seems need maintainers' approval

Done.

File: be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp (hunk 2 above)
Comment (HackToday) [reply be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp:45]: it is for two versions (avx2 and avx512) to make it run selectively according to platform features.
It was old tricky (avx2 vs non-avx2), is it ?

Comment (adonis0147) [reply be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp:45]: > it is for two versions (avx2 and avx512) to make it run selectively according to platform features. It was old tricky (avx2 vs non-avx2), is it ?

Okay, I see. I think you would better add comments like [L26](https://github.com/apache/doris/blob/ce4c8a710de456c9e919817ab7e8b59086a3726c/be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp#L26).

**Comment by HackToday** [comment]
Hi @adonis0147 do you mean two separate PRs ? One PR for upgrading bitshuffle, and another for changing BE ?
Thanks

**Comment by adonis0147** [comment]
> Hi @adonis0147 do you mean two separate PRs ? One PR for upgrading bitshuffle, and another for changing BE (until first PR merged)? Thanks

Yes.

**Comment by HackToday** [comment]
@adonis0147  The first split PR is here: https://github.com/apache/doris/pull/15993

**Comment by HackToday** [comment]
@adonis0147  hi, I am not sure the difference in Doris github run vs host run.

In my local env, I can build and run ut successfully.

The attached file is ut run in my local host.

[run-ut-test.txt](https://github.com/apache/doris/files/10441975/run-ut-test.txt)

but doris github in teamcity would complain with error

http://43.132.222.7:8111/viewLog.html?buildId=82706&buildTypeId=Doris_DorisBeUt_BeUt&tab=buildLog&_focus=260

**Comment by adonis0147** [comment]
> @adonis0147 hi, I am not sure the difference in Doris github run vs host run.
> 
> In my local env, I can build and run ut successfully.
> 
> The attached file is ut run in my local host.
> 
> [run-ut-test.txt](https://github.com/apache/doris/files/10441975/run-ut-test.txt)
> 
> but doris github in teamcity would complain with error
> 
> http://43.132.222.7:8111/viewLog.html?buildId=82706&buildTypeId=Doris_DorisBeUt_BeUt&tab=buildLog&_focus=260

Hi @HackToday , sorry to confuse you. The source of the pre-built third-party libraries which GitHub workflows use differs from the one TeamCity workflows use.

GitHub workflows use the pre-built third-party libraries from [Apache Doris Third Party Prebuilt](https://github.com/apache/doris-thirdparty/releases/tag/automation) and these libraries are up to date. However, TeamCity workflows use ones from the Docker image  [apache/doris:build-env-ldb-toolchain-latest](https://hub.docker.com/layers/apache/doris/build-env-ldb-toolchain-latest/images/sha256-fbea4c935b6664944abe118cbf961792001af5f22c0a820c2f81a0b2d8296d07?context=explore) and they are obsolete now. That is the reason why your checks failed.

In short, just wait patiently and I will let you re-trigger all these checks when everything is ready.

By the way, would you like resolve the previous comments?

File: be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp (hunk 2 above)
Comment (HackToday) [reply be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp:47]: Have resolved as comment

**Comment by HackToday** [comment]
hi @adonis0147  Have resolved. Please help check. Thanks.

File: be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp (hunk 2 above)
Comment (adonis0147) [reply be/src/olap/rowset/segment_v2/bitshuffle_wrapper.cpp:47]: Okay.

**Comment by adonis0147** [comment]
Hi @HackToday , the Docker image has been updated. Please rebase the branch on master to re-trigger the checks.

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.89 seconds
 load time: 497 seconds
 storage size: 17123413676 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230129010420_clickbench_pr_85877.html

**Comment by HackToday** [comment]
hi @adonis0147  Have rebased. Not quite sure the Po regression failed reason:
http://43.132.222.7:8111/viewLog.html?buildId=85876&tab=buildResultsDiv&buildTypeId=Doris_DorisRegression_P0Regression#testNameId-4553081828222454766

>> Caused by: java.lang.AssertionError: Expect exception msg contains 'null', but meet 'java.sql.SQLException: errCode = 2, detailMessage = There is no scanNode Backend available.[10003: not alive]'

It seems not related. How to rerun the P0 case or something else ?

**Comment by adonis0147** [comment]
> hi @adonis0147 Have rebased. Not quite sure the Po regression failed reason: http://43.132.222.7:8111/viewLog.html?buildId=85876&tab=buildResultsDiv&buildTypeId=Doris_DorisRegression_P0Regression#testNameId-4553081828222454766
> 
> > > Caused by: java.lang.AssertionError: Expect exception msg contains 'null', but meet 'java.sql.SQLException: errCode = 2, detailMessage = There is no scanNode Backend available.[10003: not alive]'
> 
> It seems not related. How to rerun the P0 case or something else ?

Hi @HackToday , these errors indicated that the workflow built the codebase successfully, but some cases failed. They are irrelevant to this PR and we are resolving these them now.

**Comment by adonis0147** [comment]
Hi @HackToday , it seems that we have resolved the issues with the workflows. Please rebase the branch on the latest master to trigger the checks again.

**Review by adonis0147 — APPROVED** [review]
LGTM

**Comment by xiaokang** [comment]
@HackToday A avx512 related bug #24145 is reported. Would you analyse and fix this problem?

**Comment by xiaokang** [comment]
Revert this PR and wait for further fix.

**Comment by HackToday** [comment]
@xiaokang sure, I have leave comment on the issue. I would double check if arch special issue or something else. Thanks

**Comment by HackToday** [comment]
@xiaokang  please help leave your comment on the issue (https://github.com/apache/doris/issues/24145) for some questions. Thanks
